File: packages/core-python/src/truewire_core/ws/rpc.py
from typing_extensions import Any, TypeVar, Generic, TypedDict
from abc import abstractmethod
from dataclasses import dataclass, field
import asyncio

from .socket import Socket

Request = TypeVar('Request', default=Any)
Reply = TypeVar('Reply', default=Any)

class Response(TypedDict, Generic[Reply]):
  id: int
  reply: Reply

@dataclass
class Rpc(Socket, Generic[Request, Reply]):
  """Multiplexed request/response client via message IDs.
  
  ### Features
  - Internal ID management
  - `rpc_request` for RPC request/response agnostic to IDs

  ### Requires implementing
  - `parse_response` to parse an incoming message into the ID and reply
  - `rpc_send` to send an identified request to the server

  ### Concurrency Contract
  1. Connection: single owner via `async with`, also supports lazy no-owner use
  2. Requests: many concurrent `rpc_request()` calls OK
  """
  replies: dict[int, asyncio.Future[Reply]] = field(default_factory=dict)
  counter: int = 0

  @abstractmethod
  def parse_response(self, msg: str | bytes) -> Response[Reply] | None:
    ...

  @abstractmethod
  async def rpc_send(self, id: int, req: Request):
    ...

  def on_msg(self, msg: str | bytes):
    response = self.parse_response(msg)
    if response is not None:
      self.replies[response['id']].set_result(response['reply'])

  async def rpc_request(self, request: Request) -> Reply:
    id = self.counter
    self.counter += 1
    self.replies[id] = asyncio.Future()
    await self.rpc_send(id, request)
    response = await self.wait(self.replies[id])
    del self.replies[id]
    return response

File: packages/core-ts/src/ws/rpc.ts
/** Multiplexed request/response client via message ids. */
import { Deferred } from './async.js'
import { Socket, type Context, type Data } from './socket.js'

/** A parsed reply: the id of the request it answers, and the reply itself. */
export interface Response<Reply> {
  id: number
  reply: Reply
}

/** The pieces of a socket `rpcRequest` needs; `Rpc` and `StreamsRpc` both provide them. */
export interface RpcHost<Request, Reply> extends Socket {
  readonly replies: Map<number, Deferred<Reply>>
  counter: number
  rpcSend(id: number, request: Request): Promise<void>
}

/** Send `request` under a fresh id and wait for its reply: the `Rpc.rpcRequest` body, shared with `StreamsRpc`. */
export async function rpcRequest<Req, Rep>(host: RpcHost<Req, Rep>, request: Req, ctx?: Context): Promise<Rep> {
  const id = host.counter++
  const reply = new Deferred<Rep>()
  host.replies.set(id, reply)
  try {
    await host.rpcSend(id, request)
    return await host.wait(reply.promise, ctx)
  } finally {
    host.replies.delete(id)
  }
}

/**
 * Multiplexed request/response client: ids are managed here, `rpcRequest` is agnostic to
 * them.
 *
 * Requires implementing `parseResponse` and `rpcSend`. Many concurrent `rpcRequest()`
 * calls are fine.
 */
export abstract class Rpc<Request = unknown, Reply = unknown> extends Socket implements RpcHost<Request, Reply> {
  readonly replies = new Map<number, Deferred<Reply>>()
  counter = 0

  /** Parse a frame into an identified reply, or `null` for frames that are not one. */
  abstract parseResponse(msg: Data): Response<Reply> | null
  /** Write `request` to the wire, tagged with `id` the way the protocol correlates replies. */
  abstract rpcSend(id: number, request: Request): Promise<void>

  override async close(): Promise<void> {
    await super.close()
    this.replies.clear()
  }

  onMsg(msg: Data): void {
    const response = this.parseResponse(msg)
    if (response !== null) this.replies.get(response.id)?.resolve(response.reply)
  }

  rpcRequest(request: Request): Promise<Reply> {
    return rpcRequest(this, request)
  }
}
